File: rock/sdk/job/result.py
"""Result models for the Job system.

Base classes: TrialResult, JobStatus, JobResult[T].
Harbor-specific subclasses in rock.sdk.agent.models.trial.result.
"""

from __future__ import annotations

import re
from datetime import datetime
from enum import Enum
from typing import Generic, TypeVar

from pydantic import BaseModel, Field

# ---------------------------------------------------------------------------
# TrialResult base — common fields
# ---------------------------------------------------------------------------


class ExceptionInfo(BaseModel):
    """General exception info."""

    exception_type: str = ""
    exception_message: str = ""
    exception_traceback: str = ""
    occurred_at: str | None = None


class TrialResult(BaseModel):
    """Base class for a single execution result — common fields.

    Harbor's TrialResult inherits this class and adds agent_info, verifier_result, etc.
    Subclasses can override the score and status properties.
    """

    task_name: str = ""
    exception_info: ExceptionInfo | None = None
    started_at: str | None = None
    finished_at: str | None = None
    # G5: process-level outputs captured by JobExecutor
    raw_output: str = ""
    exit_code: int = 0

    @property
    def score(self) -> float:
        return 0.0

    @property
    def status(self) -> str:
        return "failed" if self.exception_info else "completed"

    @property
    def duration_sec(self) -> float:
        if self.started_at and self.finished_at:
            try:
                start = datetime.fromisoformat(self.started_at.replace("Z", "+00:00"))
                end = datetime.fromisoformat(self.finished_at.replace("Z", "+00:00"))
                return (end - start).total_seconds()
            except (ValueError, TypeError):
                pass
        return 0.0


# ---------------------------------------------------------------------------
# JobStatus + JobResult[T]
# ---------------------------------------------------------------------------


class JobStatus(str, Enum):
    """Job status enum."""

    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"


T = TypeVar("T", bound=TrialResult)


class JobResult(BaseModel, Generic[T]):
    """Aggregated result of a complete job run.

    Generic over trial result type T:
      - JobResult[TrialResult]       — base (new Job system)
      - JobResult[HarborTrialResult] — Harbor agent system
    """

    job_id: str = ""
    status: JobStatus = JobStatus.COMPLETED
    labels: dict[str, str] = Field(default_factory=dict)
    trial_results: list[T] = Field(default_factory=list)
    raw_output: str = ""
    exit_code: int = 0

    @property
    def score(self) -> float:
        if not self.trial_results:
            return 0.0
        return sum(t.score for t in self.trial_results) / len(self.trial_results)

    @property
    def n_completed(self) -> int:
        return sum(1 for t in self.trial_results if t.status == "completed")

    @property
    def n_failed(self) -> int:
        return sum(1 for t in self.trial_results if t.status == "failed")


# ---------------------------------------------------------------------------
# BashTrialResult — parses score from stdout Score Summary
# ---------------------------------------------------------------------------

_SCORE_RE = re.compile(
    r"===\s*Score\s+Summary\s*===\s*\n.*?(?:score|task_score):\s*([\d.]+)",
    re.DOTALL | re.IGNORECASE,
)


class BashTrialResult(TrialResult):
    """TrialResult for Bash Jobs — parses score from stdout Score Summary block.

    BenchHub Bash Jobs emit a ``=== Score Summary ===`` block at the end of
    stdout. This subclass extracts the score so it surfaces in JobResult
    and can be reported to tracking backends.
    """

    _parsed_score: float | None = None

    @property
    def score(self) -> float:
        if self._parsed_score is not None:
            return self._parsed_score
        if not self.raw_output:
            return 0.0
        match = _SCORE_RE.search(self.raw_output)
        if match:
            try:
                self._parsed_score = float(match.group(1))
                return self._parsed_score
            except (ValueError, TypeError):
                pass
        return 0.0

File: rock/ts-sdk/src/job/result.ts
/**
 * Result models for the Job system.
 *
 * Base classes: TrialResult, JobStatus, JobResult.
 * Harbor-specific subclasses in bench/models/trial/result.ts extend these.
 */

import { z } from 'zod';

// ---------------------------------------------------------------------------
// JobStatus
// ---------------------------------------------------------------------------

/** Job status enum — matches Python rock.sdk.job.result.JobStatus. */
export enum JobStatus {
  PENDING = 'pending',
  RUNNING = 'running',
  COMPLETED = 'completed',
  FAILED = 'failed',
  CANCELLED = 'cancelled',
}

// ---------------------------------------------------------------------------
// ExceptionInfo
// ---------------------------------------------------------------------------

export const ExceptionInfoSchema = z.object({
  exception_type: z.string().default(''),
  exception_message: z.string().default(''),
  exception_traceback: z.string().default(''),
  occurred_at: z.string().nullable().default(null),
});

export type ExceptionInfo = z.infer<typeof ExceptionInfoSchema>;

// ---------------------------------------------------------------------------
// TrialResult
// ---------------------------------------------------------------------------

const _TrialResultPayloadSchema = z.object({
  task_name: z.string().default(''),
  exception_info: ExceptionInfoSchema.nullable().default(null),
  started_at: z.string().nullable().default(null),
  finished_at: z.string().nullable().default(null),
  raw_output: z.string().default(''),
  exit_code: z.number().int().default(0),
});

type _TrialResultPayload = z.infer<typeof _TrialResultPayloadSchema>;

/**
 * TrialResult — base class for a single execution result.
 *
 * Computed getters (score, status, duration_sec) are attached via a transform
 * so they work on any parsed data without a class wrapper.
 */
export type TrialResult = _TrialResultPayload & {
  readonly score: number;
  readonly status: string;
  readonly duration_sec: number;
};

function _addTrialResultComputed(payload: _TrialResultPayload): TrialResult {
  return Object.defineProperties(payload, {
    score: {
      get(): number {
        return 0.0;
      },
      enumerable: true,
      configurable: true,
    },
    status: {
      get(): string {
        return this.exception_info ? 'failed' : 'completed';
      },
      enumerable: true,
      configurable: true,
    },
    duration_sec: {
      get(): number {
        if (this.started_at && this.finished_at) {
          try {
            const start = new Date(this.started_at.replace('Z', '+00:00')).getTime();
            const end = new Date(this.finished_at.replace('Z', '+00:00')).getTime();
            return (end - start) / 1000;
          } catch {
            return 0.0;
          }
        }
        return 0.0;
      },
      enumerable: true,
      configurable: true,
    },
  }) as TrialResult;
}

/** Zod schema for TrialResult: validates data and attaches computed getters. */
export const TrialResultSchema = _TrialResultPayloadSchema.transform((payload) =>
  _addTrialResultComputed(payload)
);

// ---------------------------------------------------------------------------
// JobResult
// ---------------------------------------------------------------------------

/**
 * JobResult — aggregated result of a complete job run.
 *
 * Generic over trial result type:
 *   - JobResult<TrialResult>       — base (new Job system)
 *   - JobResult<HarborTrialResult> — Harbor agent system
 */

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function _JobResultPayloadSchema<T extends z.ZodTypeAny>(trialResultSchema: T) {
  return z.object({
    job_id: z.string().default(''),
    status: z.nativeEnum(JobStatus).default(JobStatus.COMPLETED),
    labels: z.record(z.string()).default({}),
    trial_results: z.array(trialResultSchema).default([]),
    raw_output: z.string().default(''),
    exit_code: z.number().int().default(0),
  });
}

type _JobResultPayload<T> = {
  job_id: string;
  status: JobStatus;
  labels: Record<string, string>;
  trial_results: T[];
  raw_output: string;
  exit_code: number;
};

/**
 * JobResult type — the payload plus computed getters for score, n_completed, n_failed.
 */
export type JobResult<T extends { score: number; status: string } = TrialResult> =
  _JobResultPayload<T> & {
    readonly score: number;
    readonly n_completed: number;
    readonly n_failed: number;
  };

function _addJobResultComputed<T extends { score: number; status: string }>(
  payload: _JobResultPayload<T>
): JobResult<T> {
  return Object.defineProperties(payload, {
    score: {
      get(): number {
        if (this.trial_results.length === 0) return 0.0;
        let total = 0;
        for (const t of this.trial_results) {
          total += t.score;
        }
        return total / this.trial_results.length;
      },
      enumerable: true,
      configurable: true,
    },
    n_completed: {
      get(): number {
        let n = 0;
        for (const t of this.trial_results) {
          if (t.status === 'completed') n++;
        }
        return n;
      },
      enumerable: true,
      configurable: true,
    },
    n_failed: {
      get(): number {
        let n = 0;
        for (const t of this.trial_results) {
          if (t.status === 'failed') n++;
        }
        return n;
      },
      enumerable: true,
      configurable: true,
    },
  }) as JobResult<T>;
}

/**
 * Create a JobResultSchema for a given trial result schema.
 *
 * Usage: `const schema = JobResultSchema(TrialResultSchema);`
 */
export function JobResultSchema<T extends z.ZodTypeAny>(trialResultSchema: T) {
  return _JobResultPayloadSchema(trialResultSchema).transform((payload) =>
    _addJobResultComputed(payload)
  );
}
